linden-core/src/main/java/com/xiaomi/linden/cluster/ShardClient.java

```diff
@@ -40,73 +45,93 @@
     // the path may be created later, so the listener will not work
     // and the isAvailable will always be false.
     if (!zkClient.exists(path)) {
       zkClient.createPersistent(path, true);
     }
-    host = CommonUtils.getLocalHost();
+
     List<String> children = zkClient.getChildren(path);
+    final Map<String, Map.Entry<String, LindenService.ServiceIface>> lindenClients = new ConcurrentHashMap<>();
     zkClient.subscribeChildChanges(path, new LindenZKListener(path, children) {
       @Override
       public void onChildChange(String parent, List<String> children, List<String> newAdded, List<String> deleted) {
-        buildClient(newAdded, deleted);
+        for (String node : newAdded) {
+          String fullPath = FilenameUtils.separatorsToUnix(FilenameUtils.concat(path, node));
+          byte[] bytes = zkClient.readData(fullPath);
+
+          ServiceInstance instance = JSONObject.parseObject(new String(bytes), ServiceInstance.class);
+          String hostPort = String.format("%s:%s", instance.getServiceEndpoint().getHost(),
+              instance.getServiceEndpoint().getPort());
+          if (localHostPort.equals(hostPort)) {
+            haslocalClient = true;
+            lindenClients.put(node, new AbstractMap.SimpleEntry<>(hostPort, localClient));
+            LOGGER.info("Linden local node {} {} joined shard {}.", node, hostPort, shardId);
+          } else {
+            LindenService.ServiceIface client = Thrift.newIface(hostPort, LindenService.ServiceIface.class);
+            lindenClients.put(node, new AbstractMap.SimpleEntry<>(hostPort, client));
+            LOGGER.info("Linden node {} {} joined shard {}.", node, hostPort, shardId);
+          }
+        }
+        for (String node : deleted) {
+          if (lindenClients.containsKey(node)) {
+            String hostPort = lindenClients.get(node).getKey();
+            lindenClients.remove(node);
+            LOGGER.info("Linden node {} {} left shard {}.", node, hostPort, shardId);
+          }
+        }
+
+        // ensure the new node overrides the old node.
+        List<String> sortedNodes = new ArrayList<>();
+        for (String node : lindenClients.keySet()) {
+          sortedNodes.add(node);
+        }
+        Collections.sort(sortedNodes, Collections.reverseOrder());
+
+        Set<String> uniqueClients = new HashSet<>();
+        for (String node : sortedNodes) {
+          if (lindenClients.containsKey(node)) {
+            String hostPort = lindenClients.get(node).getKey();
+            if (uniqueClients.contains(hostPort)) {
+              lindenClients.remove(node);
+              LOGGER.warn("Linden node {} {} is duplicated in shard {}, removed!", node, hostPort, shardId);
+            } else {
+              uniqueClients.add(hostPort);
+            }
+          }
+        }
+
+        LOGGER.info("{} Linden node in shard {}.", lindenClients.size(), shardId);
+        clients = new ArrayList<>();
```

Comment:
Done. Also volatile the clients field, make sure other threads will see and use the new reference as discuss here: http://stackoverflow.com/questions/1351223/thread-safe-setting-of-a-variable-java